### ISSUE TITLE
[v2] HTTP Redirects Don't Work

### ISSUE THREAD

**Issue by alexandrud** [kubernetes-sigs/aws-alb-ingress-controller#1475]
Hi, 

I'm trying to do a HTTP redirect, but the controller throws an error and all the rules end up on the port 80 listner on 2.0.0-rc2.

Example ingress configuration:
`apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  annotations:
    alb.ingress.kubernetes.io/backend-protocol: HTTP
    alb.ingress.kubernetes.io/group.name: portal
    alb.ingress.kubernetes.io/group.order: "1000"
    alb.ingress.kubernetes.io/healthcheck-interval-seconds: "30"
    alb.ingress.kubernetes.io/healthcheck-path: /
    alb.ingress.kubernetes.io/listen-ports: '[{"HTTP": 80}, {"HTTPS":443}]'
    alb.ingress.kubernetes.io/load-balancer-attributes: access_logs.s3.enabled=true,access_logs.s3.bucket=some-bucket,access_logs.s3.prefix=elb,deletion_protection.enabled=true
    alb.ingress.kubernetes.io/actions.ssl-redirect: '{"Type": "redirect", "RedirectConfig": { "Protocol": "HTTPS", "Port": "443", "StatusCode": "HTTP_301"}}'
    alb.ingress.kubernetes.io/scheme: internet-facing
    alb.ingress.kubernetes.io/ssl-policy: ELBSecurityPolicy-TLS-1-2-Ext-2018-06
    alb.ingress.kubernetes.io/tags: Environment=qa,Owner=owner
    kubernetes.io/ingress.class: alb
`

However if I do a "kubectl get ingress <name> -o yaml" I don't see the actions.ssl-redirect annotation. The error in the logs is below.

`{"level":"info","ts":1602160216.5147913,"msg":"creating listener rule","stackID":"portal","resourceID":"1"}
{"level":"info","ts":1602160216.5488174,"msg":"created listener rule","stackID":"portal","resourceID":"1","arn":"arn:aws:elasticloadbalancing:us-east-1:394023711111:listener-rule/app/k8s-portal-9a840ecc96/96deecb721f03294/c686925b2be72f71/4ba2c75f7812022a"}
{"level":"info","ts":1602160216.5488677,"msg":"creating listener rule","stackID":"portal","resourceID":"2"}
{"level":"info","ts":1602160216.5932212,"msg":"created listener rule","stackID":"portal","resourceID":"2","arn":"arn:aws:elasticloadbalancing:us-east-1:394023711111:listener-rule/app/k8s-portal-9a840ecc96/96deecb721f03294/c686925b2be72f71/a9e953033b4664f6"}
{"level":"info","ts":1602160216.5932865,"msg":"successfully deployed model"}
{"level":"error","ts":1602161667.2760198,"logger":"controller-runtime.controller","msg":"Reconciler error","controller":"ingress","name":"portal","namespace":"","error":"missing actions.ssl-redirect configuration","errorVerbose":"missing actions.ssl-redirect configuration\nsigs.k8s.io/aws-load-balancer-controller/pkg/ingress.(*defaultEnhancedBackendBuilder).buildActionViaAnnotation\n\t/workspace/pkg/ingress/enhanced_backend_builder.go:94\nsigs.k8s.io/aws-load-balancer-controller/pkg/ingress.(*defaultEnhancedBackendBuilder).Build\n\t/workspace/pkg/ingress/enhanced_backend_builder.go:57\nsigs.k8s.io/aws-load-balancer-controller/pkg/ingress.(*defaultModelBuildTask).buildListenerRules\n\t/workspace/pkg/ingress/model_build_listener_rules.go:20\nsigs.k8s.io/aws-load-balancer-controller/pkg/ingress.(*defaultModelBuildTask).run\n\t/workspace/pkg/ingress/model_builder.go:183\nsigs.k8s.io/aws-load-balancer-controller/pkg/ingress.(*defaultModelBuilder).Build\n\t/workspace/pkg/ingress/model_builder.go:104\nsigs.k8s.io/aws-load-balancer-controller/controllers/ingress.(*GroupReconciler).Reconcile\n\t/workspace/controllers/ingress/group_controller.go:85\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.1/pkg/internal/controller/controller.go:233\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.1/pkg/internal/controller/controller.go:209\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).worker\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.1/pkg/internal/controller/controller.go:188\nk8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1\n\t/go/pkg/mod/k8s.io/apimachinery@v0.18.4/pkg/util/wait/wait.go:155\nk8s.io/apimachinery/pkg/util/wait.BackoffUntil\n\t/go/pkg/mod/k8s.io/apimachinery@v0.18.4/pkg/util/wait/wait.go:156\nk8s.io/apimachinery/pkg/util/wait.JitterUntil\n\t/go/pkg/mod/k8s.io/apimachinery@v0.18.4/pkg/util/wait/wait.go:133\nk8s.io/apimachinery/pkg/util/wait.Until\n\t/go/pkg/mod/k8s.io/apimachinery@v0.18.4/pkg/util/wait/wait.go:90\nruntime.goexit\n\t/usr/local/go/src/runtime/asm_amd64.s:1374","stacktrace":"github.com/go-logr/zapr.(*zapLogger).Error\n\t/go/pkg/mod/github.com/go-logr/zapr@v0.1.0/zapr.go:128\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.1/pkg/internal/controller/controller.go:235\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.1/pkg/internal/controller/controller.go:209\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).worker\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.1/pkg/internal/controller/controller.go:188\nk8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1\n\t/go/pkg/mod/k8s.io/apimachinery@v0.18.4/pkg/util/wait/wait.go:155\nk8s.io/apimachinery/pkg/util/wait.BackoffUntil\n\t/go/pkg/mod/k8s.io/apimachinery@v0.18.4/pkg/util/wait/wait.go:156\nk8s.io/apimachinery/pkg/util/wait.JitterUntil\n\t/go/pkg/mod/k8s.io/apimachinery@v0.18.4/pkg/util/wait/wait.go:133\nk8s.io/apimachinery/pkg/util/wait.Until\n\t/go/pkg/mod/k8s.io/apimachinery@v0.18.4/pkg/util/wait/wait.go:90"}`

**Comment by M00nF1sh** [comment]
@alexandrud would you help paste your whole ingress spec?
also, seems you have multiple ingresses in cluster, is there anyone used `ssl-redirect` as serviceName but didn't configure the corresponding `actions.ssl-redirect` annotation?

**Comment by alexandrud** [comment]
Hi,

Bellow are the outputs for the helm chart and the ingress object on the cluster. I've redacted a couple of details.

There are 2 other ingresses I'm testing. They have a ssl-redirect action annotation, but no ssl-redirect service. I'm going to remove the ssl-redirect annotation and if anything changes I'll post here.

But on another note I noticed another thing which might shed some light on what's happening:
- setup listen ports 80 and 443 from annotations (removed redirect annotation)
- setup at least 2 routes (removed redirect block)

The routes get set on port 80, but not port 443.

Thanks!

This is the output of "kubectl get ingress portal-core -o yaml"

```apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  annotations:
    alb.ingress.kubernetes.io/actions.redirect: '{"Type": "redirect", "RedirectConfig":
      { "Protocol": "HTTPS", "Port": "443", "StatusCode": "HTTP_301"}}'
    alb.ingress.kubernetes.io/backend-protocol: HTTPS
    alb.ingress.kubernetes.io/group.name: portal
    alb.ingress.kubernetes.io/group.order: "1000"
    alb.ingress.kubernetes.io/healthcheck-interval-seconds: "30"
    alb.ingress.kubernetes.io/healthcheck-path: /
    alb.ingress.kubernetes.io/healthcheck-protocol: HTTPS
    alb.ingress.kubernetes.io/listen-ports: '[{"HTTP": 80}, {"HTTPS":443}]'
    alb.ingress.kubernetes.io/load-balancer-attributes: access_logs.s3.enabled=true,access_logs.s3.bucket=qa_bucket,access_logs.s3.prefix=elb,deletion_protection.enabled=true
    alb.ingress.kubernetes.io/scheme: internet-facing
    alb.ingress.kubernetes.io/ssl-policy: ELBSecurityPolicy-TLS-1-2-Ext-2018-06
    alb.ingress.kubernetes.io/tags: Environment=qa,Owner=owner
    kubernetes.io/ingress.class: alb
    meta.helm.sh/release-name: portal-core
    meta.helm.sh/release-namespace: aep-helm-tests
  creationTimestamp: "2020-10-08T16:20:09Z"
  finalizers:
  - alb.ingress.k8s.aws/portal
  generation: 3
  labels:
    aep/component: portal
    app.kubernetes.io/instance: portal-core
    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/name: portal-core
    app.kubernetes.io/version: 2.6.0
    helm.sh/chart: portal-core-0.1.0
  name: portal-core
  namespace: aep-helm-tests
  resourceVersion: "231837313"
  selfLink: /apis/extensions/v1beta1/namespaces/aep-helm-tests/ingresses/portal-core
  uid: 3f537765-45cf-4552-a06d-886b89adfe1e
spec:
  rules:
  - host: redacted
    http:
      paths:
      - backend:
          serviceName: ssl-redirect
          servicePort: use-annotation
        path: /*
      - backend:
          serviceName: portal-core
          servicePort: 30080
        path: /*
status:
  loadBalancer:
    ingress:
    - hostname: redacted```

This is the output of the helm chart:

```
# Source: portal-core/templates/ingress.yaml
apiVersion: networking.k8s.io/v1beta1
kind: Ingress
metadata:
  name: portal-core
  labels:
    helm.sh/chart: portal-core-0.1.0
    app.kubernetes.io/name: portal-core
    app.kubernetes.io/instance: portal-core
    app.kubernetes.io/version: 2.6.0
    app.kubernetes.io/managed-by: Helm
    aep/component: portal
  annotations:
    alb.ingress.kubernetes.io/tags: Environment=qa,Owner=owner
    alb.ingress.kubernetes.io/group.name: portal
    alb.ingress.kubernetes.io/healthcheck-path: /
    alb.ingress.kubernetes.io/load-balancer-attributes: access_logs.s3.enabled=true,access_logs.s3.bucket=qa_bucket,access_logs.s3.prefix=elb,deletion_protection.enabled=true
    alb.ingress.kubernetes.io/actions.redirect: '{"Type": "redirect", "RedirectConfig":
      { "Protocol": "HTTPS", "Port": "443", "StatusCode": "HTTP_301"}}'
    alb.ingress.kubernetes.io/backend-protocol: HTTPS
    alb.ingress.kubernetes.io/group.order: "1000"
    alb.ingress.kubernetes.io/healthcheck-interval-seconds: "30"
    alb.ingress.kubernetes.io/healthcheck-protocol: HTTPS
    alb.ingress.kubernetes.io/listen-ports: '[{"HTTP": 80}, {"HTTPS":443}]'
    alb.ingress.kubernetes.io/scheme: internet-facing
    alb.ingress.kubernetes.io/ssl-policy: ELBSecurityPolicy-TLS-1-2-Ext-2018-06
    kubernetes.io/ingress.class: alb
spec:
  rules:
    - host: portal-qa-directvcommercial.test.att.com
      http:
        paths:
          - path: /*
            backend:
              serviceName: ssl-redirect
              servicePort: use-annotation
          - path: /*
            backend:
              serviceName: portal-core
              servicePort: 30080
```

**Comment by M00nF1sh** [comment]
closing this since it's addressed by [v2.0.0-rc3 release](https://github.com/kubernetes-sigs/aws-alb-ingress-controller/releases/tag/v2.0.0-rc3) :D.
feel free to reopen if further issue exists.
